server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/data/StarTreeDataWriter.java

```diff
@@ -66,21 +66,32 @@ private static void writeStarTreeNodes(IndexOutput output, InMemoryTreeNode root
         while (!queue.isEmpty()) {
             InMemoryTreeNode node = queue.remove();
 
-            if (node.children == null || node.children.isEmpty()) {
+            if ((node.children == null || node.children.isEmpty()) && node.childStarNode == null) {
                 writeStarTreeNode(output, node, ALL, ALL);
             } else {
 
-                // Sort all children nodes based on dimension value
-                List<InMemoryTreeNode> sortedChildren = new ArrayList<>(node.children.values());
-                sortedChildren.sort(
-                    Comparator.comparingInt(InMemoryTreeNode::getNodeType).thenComparingLong(InMemoryTreeNode::getDimensionValue)
-                );
-
+                int totalNumberOfChildren = 0;
                 int firstChildId = currentNodeId + queue.size() + 1;
-                int lastChildId = firstChildId + sortedChildren.size() - 1;
+
+                if (node.childStarNode != null) {
+                    totalNumberOfChildren++;
+                    queue.add(node.childStarNode);
+                }
+
+                if (node.children != null) {
+                    // Sort all children nodes based on dimension value
+                    List<InMemoryTreeNode> sortedChildren = new ArrayList<>(node.children.values());
+                    sortedChildren.sort(
+                        Comparator.comparingInt(InMemoryTreeNode::getNodeType).thenComparingLong(InMemoryTreeNode::getDimensionValue)
```

Comment:
Why are we not removing this sort btw ?

LinkedHashMap + removing the sorting in tests as part of validateFileFormats pretty much fixes all the tests

Reply:
removed sorting logic